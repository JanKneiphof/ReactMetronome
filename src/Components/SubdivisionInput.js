import { Button, ButtonGroup, Card, Grid, Typography } from '@material-ui/core';
import React, { Component } from 'react';
import {captionInCard, gridInCard} from '../styles/card.styles'


class SubdivisionInput extends Component {

    handleClick(index) {
        this.props.changeSubdivision(index)
    }

    createButtonGroups(size) {
        let buttons = []

        for (let index = 1; index <= size; index++) {
            let intensity = this.props.beatAccentuation.get(index - 1)
            let accentString = "â–ˆ".repeat(intensity)
            if (accentString.length === 0) {
                accentString = "mute"
            }
            buttons.push(
                <Grid key={index} item>
                    <Grid container alignItems="center" direction="column">
                        <Grid item>
                            <Typography>{accentString}</Typography>
                        </Grid>
                        <Grid item>
                            <ButtonGroup
                                key={"upTo" + index}
                                size="small"
                            >
                                <Button
                                    size="small"
                                    onClick={this.props.changeSubdivision.bind(this, index)}
                                    variant="outlined"
                                    value={intensity}
                                    aria-label={"Subdivision " + index}
                                >
                                    {index}
                                </Button>
                            </ButtonGroup>
                        </Grid>
                    </Grid>
                </Grid>
            )

        }
        return (<Grid container style={gridInCard} direction="row">{buttons}</Grid>)
    }

    render() {
        return (
            <Card variant="outlined">
                <Typography variant="caption" style={captionInCard}>Beat Accents</Typography>
                {this.createButtonGroups(this.props.numberOfSubdivisions)}
            </Card>
        )
    }
}

export default SubdivisionInput