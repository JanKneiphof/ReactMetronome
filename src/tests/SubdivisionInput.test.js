import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import SubdivisionInput from '../Components/SubdivisionInput'
import { getAllByLabelText } from '@testing-library/react';

let container = document.createElement("div");
beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
});
afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

describe("SubdivisionInput", () => {
    test("matches Snapshot", () => {
        act(() => {
            ReactDOM.render(<SubdivisionInput changeSubdivision={() => {}} beatAccentuation={new Map([[0,3],[1,1],[2,1],[3,1]])} numberOfSubdivisions="4"></SubdivisionInput>, container)
        });
        expect(container.firstChild).toMatchSnapshot()
    })

    test("renders correct amount of Buttons", () => {
        let correctAmount = 4
        act(() => {
            ReactDOM.render(<SubdivisionInput changeSubdivision={() => {}} beatAccentuation={new Map()} numberOfSubdivisions={correctAmount}></SubdivisionInput>, container)
        })
        expect(getAllByLabelText(container, /Subdivision \d/).length).toBe(correctAmount)
    })
    test("renders correct amount of Buttons, even with big numbers", () => {
        let correctAmount = 123
        act(() => {
            ReactDOM.render(<SubdivisionInput changeSubdivision={() => {}} beatAccentuation={new Map()} numberOfSubdivisions={correctAmount}></SubdivisionInput>, container)
        })
        expect(getAllByLabelText(container, /Subdivision \d/).length).toBe(correctAmount)
    })
})