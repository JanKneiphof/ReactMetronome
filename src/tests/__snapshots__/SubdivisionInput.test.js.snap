// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubdivisionInput matches Snapshot 1`] = `
<div
  class="MuiGrid-root MuiGrid-container"
>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
    >
      <span
        class="MuiTypography-root MuiTypography-caption"
        style="margin: 10px;"
      >
        Beat Accents
      </span>
      <div
        class="MuiGrid-root MuiGrid-container"
        style="margin: 10px;"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                ███
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiButtonGroup-root"
                role="group"
              >
                <button
                  aria-label="Subdivision 1"
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                  value="1"
                >
                  <span
                    class="MuiButton-label"
                  >
                    1
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                █
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiButtonGroup-root"
                role="group"
              >
                <button
                  aria-label="Subdivision 2"
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                  value="2"
                >
                  <span
                    class="MuiButton-label"
                  >
                    2
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                █
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiButtonGroup-root"
                role="group"
              >
                <button
                  aria-label="Subdivision 3"
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                  value="3"
                >
                  <span
                    class="MuiButton-label"
                  >
                    3
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                █
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiButtonGroup-root"
                role="group"
              >
                <button
                  aria-label="Subdivision 4"
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
                  tabindex="0"
                  type="button"
                  value="4"
                >
                  <span
                    class="MuiButton-label"
                  >
                    4
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
